stdpopsim/models.py

```diff
@@ -137,6 +137,153 @@ def asdict(self):
+
+        `num_steps` parameter will determine how many points along the
+        time axis are returned.
+
+        Returns a tuple where the first value
+        is the steps (in generations) and the second is a numpy
+        ndarray that will contain the coalescence rate for
+        the entire population.
+        """
+
+        num_pops = self.num_pops()
+        assert(len(num_samples) == num_pops)
+        P = np.zeros([num_pops**2, num_pops**2])
+        index_array = np.array(range(num_pops**2)).reshape([num_pops, num_pops])
+        for x in range(num_pops):
+            for y in range(num_pops):
+                K_x = num_samples[x]
+                K_y = num_samples[y]
+                P[index_array[x, y], index_array[x, y]] = K_x * (K_y - (x == y))
+        P = P / np.sum(P)
+        r = np.zeros(num_steps)
+        steps = np.linspace(0, end, num_steps)
+        dt = steps[1] - steps[0]
+        mass_migration_objects = []
+        mass_migration_steps = []
+        for demo in self.demographic_events:
+            if type(demo) == msprime.simulations.MassMigration:
+                mass_migration_objects.append(demo)
+                mass_migration_steps.append(int(demo.time/dt))
+        for j in range(num_steps):
+            N_M = self.pop_size_and_migration_at_t(steps[j])
+            N = N_M["population_sizes"]
+            M = N_M["migration_matrix"]
+            C = np.zeros([num_pops**2, num_pops**2])
+            index_array = np.array(range(num_pops**2)).reshape([num_pops, num_pops])
+            for idx in range(num_pops):
+                C[index_array[idx, idx], index_array[idx, idx]] = 1 / (2 * N[idx])
+            for idx, row in enumerate(M):
+                M[idx][idx] = -1 * sum(row)
+            Identity = np.eye(len(M[0]))
+            G = (np.kron(M, Identity) + np.kron(Identity, M)) - C
+            if j in mass_migration_steps:
+                idx = mass_migration_steps.index(j)
+                ds = mass_migration_objects[idx].time - (j * dt)
+                a = mass_migration_objects[idx].source
+                b = mass_migration_objects[idx].dest
+                p = mass_migration_objects[idx].proportion
+                J = np.array(range(num_pops**2)).reshape([num_pops, num_pops])
+                S = np.eye(num_pops**2, num_pops**2)
+                for x in range(num_pops):
+                    if x == a:
+                        S[J[a, a], J[a, b]] = S[J[a, a], J[b, a]] = p * (1 - p)
+                        S[J[a, a], J[b, b]] = p ** 2
+                    else:
+                        S[J[x, a], J[x, b]] = S[J[a, x], J[b, x]] = p
+                        S[J[x, a], J[x, a]] -= p
+                        S[J[a, x], J[a, x]] -= p
+                P = np.matmul(P, S)
+                P = np.matmul(P, scipy.linalg.expm((dt - ds) * G))
+            else:
+                P = np.matmul(P, scipy.linalg.expm(dt * G))
+            r[j] = np.sum(np.matmul(P, C)) / np.sum(P)
```

Comment:
@petrelharp I keep getting `RuntimeWarning: invalid value encountered in double_scalars
  r[j] = np.sum(np.matmul(P, C)) / np.sum(P)` for this line when running on a model with mass migration. Shall we check that np.sum(P) != 0 before dividing?

Reply:
It should never be the case that `np.sum(P)` is zero. Can you check where the warning is coming from? Is it really from the `1/np.sum(P)`? Is it because of a `nan` somewhere? Or what?

Reply:
No, It is because of the np.sum(P) == 0, (I checked with an assertion). I'll investigate more now.

Reply:
A good sanity check while debugging would be to set C to zero (or, just have very, very large population sizes) and verify that the rows of `P` all sum to 1.

Reply:
This should be true for every time step? There must be a bug somewhere because all the rows are summing close to 0.1 when C is kept as zeros for each iteration.

Reply:
Sorry, I meant that the sum of *all* of `P` should be 1 at all times if coalescence rates are zero.

Reply:
At all times, sum(P) should be the probability that the lineages of a randomly chosen pair of samples have not yet coalesced, so at time `t`, if all population sizes are greater than `N`, we should always have that `1 - t / (2*N) <= sum(P) <= 1`.

Reply:
Looks like my head cold is preventing me from thinking about this too clearly, but I can say that P converges to 0 with any model if the `end` time is far enough.

Reply:
So this should never be a problem so long as `end` isn't too far being when the species demographic defines to start. Maybe we should just limit how large `end` can be?

Reply:
because if you go back to far, the probability that the lineages of a randomly chosen pair of samples have not yet coalesced converges to zero, no?

Reply:
My first suspicion is that this is a bug, because `end` would have to be really, really big for this to be a legitimate problem.

Reply:
Hmm, doesn't seem to be the case in the models I have tried - for Gutenkunst, P converges around 9500 years ago